Code Insights: [GQL API] Add first naive version of new GQL Code Insights API

vovakulikov opened the pull request:
Related to https://github.com/sourcegraph/sourcegraph/issues/33933

# Background

I'm just trying to develop a more straightforward API than our existing one in this PR. By this I mean going away from chart type abstraction in our mutations (createLineChart, createBarChart, cratePieChart ...etc) and moving to mutation that are based on our domain insight model (create<SearchInsight | Capture group | Lang stats | ... > mutations) 

This PR aims to solve only the insight entity part in our GQL API. The dashboard model is outside of the scope of this PR. This should be used for informational purposes only and should not be used or merged anywhere in the backend codebase.

## Data types

So, in a nutshell, in this PR schema, I split all insights that we have into two categories 
- Series like (data that represents points through time) 
- Categorical like (data that doesn't have any timeline-like information and has a clear set of categories and their values) 

After that, we can express all possible insights that we currently have through those types. 
- Capture group insight - series-like
- Search based insight - series-like
- Lang stats insight - categorical like
- Compute? - categorical like 

If we need to add a new type of insight - let's say treemap we won't change anything about the existing API and extend our API with another type. See `NewInsight` union type in this PR schema. 

## Chart types

We can determine a set of visual settings for the insight model based on data types. This means that the data type determines the set of possible chart types. 

- Series-like data - line chart, bar chart
- Categorical-like data - pie chart, bar chart (but with ordinary x axis), bubble chart ...

By this, we force only possible chart types for insight data that we have in our API. For example, it doesn't make sense to have Pie Chart for series-like data. (In theory, we could introduce another presentation chart type as a series of pie chart snapshots, but this is out of this PR's scope) 

Another good thing is in this system, it is possible to have a strict set of chart type settings in GQL types because charts from one group should have the same amount of settings. [See related slack message about this](https://sourcegraph.slack.com/archives/C01RVEVPWLC/p1652855353130719)

**From this slack message**
> From my understanding, if we have data that goes through time (series-like data),
there is no difference between different chart types in terms of their visual settings.

> For example, if we have a line chart and we change it to a bar chart, then
- line colour becomes bar colour
- stacked line setting (when we stacked one line on another) becomes a stacked bar
- start from zero setting is also applicable for the bar (maybe not, not sure, maybe it always should be active for the bar chart)

> Same thing for categorical line chart? Let's say we change the pie chart to the bar chart
- pie arc colour becomes bar colour 
- ... something else?

Also, here is another issue in which we may want to introduce new chart/insight settings https://github.com/sourcegraph/sourcegraph/issues/33327

## Mutations

Well, here, things have been simplified dramatically. Data types and chart types are related to feature details and the settings world. 

The insight model itself comes from the domain model of our product. So I decided to pick this abstraction for the mutations rather than implement mutation with data-type and chart-type attributes because it makes more sense when creating entities and matches our domain better, for example.

We have different pages for each insight type (search-based, capture group, lang-stats), which means we have a clear separation between insight types, and this is good because each type has a different amount of domain settings. 

So mutations look like 
```gql
createSearchInsight(input: SearchInsightInput): SeriesLikeInsight
createCaptureGroupInsight(input: CaptureGroupInsightInput): SeriesLikeInsight
createLangStatsInsight(input: LangStatsInsightInput): CategoricalLikeInsight

updateSearchBasedInsight(input: SearchInsightInput): SeriesLikeInsight
updateCaptureGroupInsight(input: CaptureGroupInsightInput): SeriesLikeInsight
updateLangStatsInsight(input: LangStatsInsightInput): CategoricalLikeInsight
```

`SearchInsightInput`, `CaptureGroupInsightInput`, `LangStatsInsightInput` contain business and view logic, in output we have data-type models.

chwarwick commented:
@vovakulikov wouldn't the `categorical like` items become `series like` if we ran the same generation logic more than once  saved a timestamp with it and the `series like` become `categorical` if we only looked at the data from a single time?

I think with this schema you are coupling the way in which the values are found (search/capture group/compute/lang stats etc..) with the results will be displayed (line chart/bar chart/etc..).  Why can't I put a capture group in a pie chart if I am most interested in the current state.

I like the mutations by how the data is generated, I wonder if there is a way there to keep the same general concept but also include in the input some display options that would allow a return of a union of LineChartInsight | PieChartInsight | BarChartInsight etc.. 

I need to look though the other proposals as well but thanks for putting this together.

vovakulikov commented:
@chwarwick, thanks for a quick reply on this. 

>wouldn't the categorical like items become series like if we ran the same generation logic more than once and saved timestamp with it, and the series like become categorical if we only looked at the data from a single time?

This is 100% correct. And I like how one data type converts to other in this system. 

>Why can't I put a capture group in a pie chart if I am most interested in the current state

You can, actually. But for this, we just should have a separate API query right? (I mentioned that this is out of the scope of this PR, but here is what it may look like) 

```gql
type SeriesLikeInsight implements BaseInsight & Node {
     id: ID!
     title: String!
     series: [NewInsightSeries!]!
     seriesSliceSnapshot(date: DateTime!): [NewInsightCategoricalDatum!]!
 }
```

As you can see, we rely on the same basis here in the `seriesSliceSnapshot` query. After this, we could render `NewInsightCategoricalDatum` with a bar or pie chart it doesn't matter. 

The main purpose of its schema is to move away from chart details and concentrate our effort on data shapes and how we collect data because I think we should aim to have the following hierarchy 
Data → Chart type (derived) → Chart settings (derived)

We currently have a system where charts control data and we all can agree this doesn't work well and doesn't scale either.

CristinaBirkel commented on cmd/frontend/graphqlbackend/new-insights.graphql:
```diff
@@ -0,0 +1,161 @@
+extend type Mutation {
+    createSearchInsight(input: SearchInsightInput): SeriesLikeInsight
+    createCaptureGroupInsight(input: CaptureGroupInsightInput): SeriesLikeInsight
+    createLangStatsInsight(input: LangStatsInsightInput): CategoricalLikeInsight
```
Would this make it impossible to have multiple data series types on the same chart? (Like thinking about something like one webhook series next to a regular search series on the same chart.) I do agree this aligns well with how the webapp works right now though and makes a lot of sense from that perspective.

vovakulikov replied:
@CristinaBirkel, I noted this problem as well, but this problem comes from the fact that GQL spec doesn't allow a union type over input types. So any API that we picked would have this problem essentially. I would the best compromise that we could do here is to have a wide option fields inpyt type. (It doesn't mean that we have this optional arbitrary fields in the query API)  

```gql 
input SearchInsightSeriesInput {
     id: ID
     query: String!
     title: String!
     color: String!
     chartType: SeriesLikeChartType
     repositoryScope: RepositoryScopeInput!
     timeScope: TimeScopeInput!
     webhook: SeriesWebhookConfig
 }
 
 input SeriesWebhookConfig {
  url: String
 }
```

coury-clark replied:
I really don't think embedding other series types (webhook) into this series type (search) input makes any sense. If anything this suggests there should be a higher level object with optional inputs below it. Soon the `oneOf` [operator will likely be available in GraphQL](https://stepzen.com/blog/coming-soon-to-graphql-the-oneof-input-object) and will support this.

vovakulikov replied:
Well even better in this case if we had `oneOf` decorator we would implement this in a cleaner way. 

```gql
nput SearchInsightSeriesInput {
     id: ID
     title: String!
     color: String!
     chartType: SeriesLikeChartType
     repositoryScope: RepositoryScopeInput!
     timeScope: TimeScopeInput!
     dataSource: SeriesDataSource
 }

input SeriesDataSource @oneOf {
     query: String
     webhook: SeriesWebhookConfig
}

input SeriesWebhookConfig {
  url: String
}
```

coury-clark replied:
I agree, something along that line is likely going to be the route forward regardless of the other data shapes involved. We don't technically need to wait for `oneOf`, we can expect that to land with nullable options in an input type and add it later to enforce the invariant.

CristinaBirkel commented on cmd/frontend/graphqlbackend/new-insights.graphql:
```diff
@@ -0,0 +1,161 @@
+    title: String!
+    color: String!
+    query: String!
+    chartType: SeriesLikeChartType
```
Having the chart type be per-series is a significant change to the data model on the backend. I would probably push back against that until we know we need to support it.

vovakulikov replied:
@CristinaBirkel, I agree. But I would advocate putting this field on the series level similar that we have repository scope and time scope on the series level, but we don't use it (we don't have series with different repositories or time scope values) 

This would allow us to implement different chart types per series letter without GQL schema change in the future.

coury-clark commented:
Something to consider here is that the insight "types" we currently have defined are not nearly as solidified at the product level as this proposal implies. It's probably more applicable to think of these as _templates_ instead of _types_. An example of where this is may be a problem is this:
> Compute? - categorical like

It's not accurate to say the data generated from compute will be categorical. This data could be generated as a time series. We may also even think of very specific use cases as templates. Consider something along the lines of:

> Compute - group by author
> Compute - group by path

Both of these will use the same underlying data generation scheme. We should only think of these as templates in how we prepopulate the UI fields rather than discrete insight types.

vovakulikov commented:
@coury-clark I may agree with your statement 
> are not nearly as solidified at the product level as this proposal implies

But all this templates have a different domain settings for creation (some of them have series query, some of them have just one query on the top level, some of them don't have any series at all). I think that unifying all insights "types" under one super flexible API is a mistake, because 

1. We don't have this flexibility in our UI (each template insight type has separate UI page)
2. This API would be too low-level I think, having our internal DSL won't make our API simpler and decently increases complexity for external consumers of our API. 

> It's not accurate to say the data generated from compute will be categorical

And this is fine) As long as we have a strict query basis categorical like, series like we can present any type of data that we have generated regardless of domain insight type (compute, search based, ...)

coury-clark commented:
> As long as we have a strict query basis categorical like

I'm not entirely following what you mean by this, could you elaborate? 

> I think that unifying all insights "types" under one super flexible API is a mistake, because

I think I agree with this in principle, with a few caveats:
1. Our read path should be generalizable based on the shape of the chart. If something is supposed to render as a line chart, it shouldn't matter what generated the data. The shape should match what is expected for a line chart. Same idea applies to any chart type.
2. Input types should be domain specific. I think we might consider "Sourcegraph" a valid type here, given that all of the insights we are talking about currently live within the domain of a Sourcegraph query. A future world with webhooks (or some other external data) might have completely different invariants. It would be a mistake to try and jam them all in the same.

I think #1 above is where this proposal loses me a bit. I'm not seeing the value necessarily of the `series-like` and `category-like` types over defining schemas and shapes based on the visualization styles themselves.

vovakulikov commented:
@coury-clark 
>I'm not entirely following what you mean by this, could you elaborate?

Yeah, I think you expressed this perfectly in your last message in this GH issue thread. 
>Our read path should be generalizable based on the shape of the chart. If something is supposed to render as a line chart, it shouldn't matter what generated the data. The shape should match what is expected for a line chart. Same idea applies to any chart type.

I think this is perfectly describes this PR's idea, with one correction that some subset of chart types are not needed for special type on the backend in GQL query. This means that we could reuse the same data-shape for different types of chart. As I described in this PR description there is no difference between line type and bar type charts (as long as they both represent series like data). This is why I think that the real separation should live on the data level, and the chart type here would be just another setting field that specifies chart type (on series or on chart level)

vovakulikov commented:
@sourcegraph/code-insights-backend @sourcegraph/code-insights-frontend I added a few additional separations in query types and aligned it a bit with how our mutation work in this PR's proposition, please take a look at this and share what you think about it. 

In the nutshell 
- I'm trying to reduce FE/consumer mapping between data and its presentations. I'm doing it because this will allow us to use GQL types on the frontend directly and remove our own types. We currently have custom FE models it because our GQL API doesn't provide a good way to work with all data that we have (we always have to somehow process data on the client). 
- I added a few types of separation into the query type
    - by insight types, because this is the only way I found to avoid performance problems with the capture group insight and its series settings (like global query for generating series, repositories field, timeInterval, ...etc.
    - by data shape, because sometimes (for example on the dashboard page) we care about insight data and don't care about insight metadata, so it simplifies clients of our GQL API a lot. 
    
For better understanding, I provided a few examples of real GQL queries that we want to have on the frontend https://github.com/sourcegraph/sourcegraph/pull/35623/commits/7cd1f01f2fd9b3ee2b2b385ef69a26a77d1b6560

@unclejustin

coury-clark commented on cmd/frontend/graphqlbackend/new-insights.graphql:
```diff
@@ -0,0 +1,208 @@
+input NewInsightViewFiltersInput {
+    includeRepoRegex: String
+    excludeRepoRegex: String
+    searchContexts: [ID!]
```
It may be problematic to associate these by ID instead of a soft reference name. This is because contexts can change state in ways that does not propagate to insights (edit, delete, etc). I don't have a concrete answer, this will require some thought.

coury-clark commented on cmd/frontend/graphqlbackend/new-insights.graphql:
```diff
@@ -0,0 +1,208 @@
+    query: String!
+    title: String!
+    color: String!
+    chartType: SeriesLikeChartType
```
I don't understand how chart type can be related to the series instead of the top level insight. What if two series on the same insight have conflicting chart types?

vovakulikov replied:
@coury-clark They can't have conflicting chart types because we explicitly say that you can apply only `SeriesLikeChartType`. This is why we have this enum here.

coury-clark replied:
But each series could have a different chart type which wouldn't make any sense. I'm not talking about the possible values this field can take, I'm talking about multiple series associated with one chart.

vovakulikov replied:
@coury-clark I don't think I understand you correctly. Different types of charts for each series make a series as long those types are comparable to each other. For example, we could have one series as a line and one series as a bar. 

Of course, we can't have a pie chart and line chart on the same chart but our GQL type system guards this by allowing only comparable chart types for each series.

vovakulikov replied:
@coury-clark, this is a valid case, and I think we already have some ideas about why we will need this at some point. Here is a related comment from Alicja 
https://github.com/sourcegraph/sourcegraph/issues/33327#issuecomment-1087282481


![](https://user-images.githubusercontent.com/20326070/161507345-05f166f4-39a3-4557-8286-2435bcdc435c.png)

vovakulikov replied:
This is actually one of the reasons why I think that bounding mutations with chart types is a bad idea because I can see that this wouldn't allow us to manage all important cases in terms of data visualization like having two different but comparable chart types per series.

coury-clark replied:
I think we should consider a "combination chart" as a separate type for something like this, and require that insights relate to chart types 1-1. 

Additionally, I think this concept (or the same combo chart) can solve the problem of different series types landing under a common chart. With this rule in place insights will relate 1-1 with both series types and presentation types, but other chart types would exist to compose new charts out of existing data. This would unblock something like this proposal if we could enforce this constraint.

coury-clark replied:
Just as an example, this is how google sheets does it as well (the data there is rows / cols, we can think of cols = series and rows = data points)

<img width="596" alt="image" src="https://user-images.githubusercontent.com/5090588/173427433-cc10d2ca-da48-49ce-8f67-7145860749ca.png">

coury-clark commented on cmd/frontend/graphqlbackend/new-insights.graphql:
```diff
@@ -0,0 +1,208 @@
+# simplify data querying in consumers.
+#
+# See client/web/src/enterprise/insights/new-insights-gql-examples.ts
+interface SeriesLikeData {
```
I'm starting to understand the vision behind this interface a bit more now that I see the `seriesSlice` field. One thought that I still have is that this proposal seems to be coupling the capability (the type of response) to the insight type (which we currently loosely define in the UI, for example capture group type). Why should a capture group insight be restricted to a `series-like` response? A categorical response is equally valid (if we inspect a single point). 

Likewise, what if we want to expand language stats to support series-like?

vovakulikov replied:
@coury-clark 

> this proposal seems to be coupling the capability (the type of response) to the insight typ

This is absolutely correct, I'd tried to get along with generic forms at the beginning but I couldn't find any generic solution on how to cover all our cases. For example,
- in capture group insight we don't specify series but we get series-like results.
- series definitions (even without points) act as data itself in case of capture group insight, so we can't get access to it immediately (as we do in search-based insight) BE should collect series first that takes time. 

So we naturally have different capabilities of response based on the insight type. At least this is what it looks like from the consumer's point of view. 


>Why should a capture group insight be restricted to a series-like response? A categorical response is equally valid (if we inspect a single point)

Agree again, as you noted in this API I tried to not limit ourselves by extending series-like data with slice query which returns a categorical response. 

> Likewise, what if we want to expand language stats to support series-like?

I think in this case we just need to extend LangStatsInsight on the backend with the SeriesLikeData interface. Do you see problems with this way of extension?

vovakulikov replied:
I just want to emphasize that from the consumer (frontend) point of view we have to have a two-dimension separation between insights. We should distinct insight by type because this is how we pick an appropriate creation UI for viewing insight configuration and we have to have separation by data shape to be able to pass this data to a valid chart on the frontend. 

And this is where all complexity comes from. From my point of view, BE wants to have a low-level API where we don't have all those separations but we are building models with small pieces of low-level abstractions (I also like this approach) but FE (our main and only one consumer for now) can't have it because we have a clear separation in our product between insight types and between different type of charts. 

And we will be in the same situation as we have now where FE has to implement our own layer of abstraction where we absorb low-level abstraction complexity and build our own models. @unclejustin had a point about it why this is bad and that we have to solve this if we want to avoid gigantic problems on the FE in future.  

I hope this would help to be a little bit closer to FE's expectations about our API.

cc @CristinaBirkel

vovakulikov commented on cmd/frontend/graphqlbackend/new-insights.graphql:
```diff
@@ -0,0 +1,208 @@
+    id: ID!
+    title: String!
+    isFrozen: Boolean!
+    filters: NewInsightFilters
```
@coury-clark I think you're right about filters are not generic across all data/insight types. I think we need to figure out in which dimension filter are different - per insight type (means filters for the search based insight may be different from filters for the capture group insight) or per data type (filters are different for series like and categorical like datasets)

vovakulikov commented:
Thanks everyone 

![image](https://github.com/sourcegraph/sourcegraph/assets/18492575/b4cd2894-5d41-4407-a287-119c556a5f90)
